### PR TITLE
[SPARK-43895] Basic Repository Layout

### PR DESCRIPTION
**Pull request by grundprinzip**
## Proposed Changes

In order to prepare the implementation of the Go client. This patch prepares the repository layout.

* Adds the git submodule for the Spark repository, currently pointing at `master` for the reference to the proto files.
* Adds the Makefile and the necessary outline for the build system
* Adds the `buf` generation logic for the protos.

## Testing

`make gen && make check && make test`

### CODE REVIEW

**Comment by zhengruifeng** [comment]
Is it possible to make `apache/spark` a submodule so that we don't need to copy the proto files?

**Comment by grundprinzip** [comment]
Actually, I was looking at the submodule and found it too clunky, tried again and it's working quite nice, will change.

**Comment by grundprinzip** [comment]
@zhengruifeng @HyukjinKwon this PR should be good now and the last PR before we can bring in the changes of Bo for the base client. The next step would be adding the workflow integration.

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+clean:
+	@echo -n ">> CLEAN"
+	@$(GO) clean -i ./...
+	@rm -rf ./internal/generated
+	@rm  -f ./internal/generated.out
+	@rm -f ./coverage-all.html
+	@rm -f ./coverage-all.out
+	@rm -f ./coverage.out
+	@find . -type f -name "coverage.out" -delete
+	@printf '%s\n' '$(OK)'
```
Comment (HyukjinKwon) [inline Makefile:105]:
```suggestion
	@printf '%s\n' '$(OK)'

```

File: .gitignore [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# All generated files
+internal/generated
+internal/generated.out
+
+# Ignore Coverage Files
+coverage*
```
Comment (HyukjinKwon) [inline .gitignore:23]:
```suggestion
coverage*

```

**Review by HyukjinKwon — APPROVED** [review]

File: go.mod [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+	github.com/zeebo/xxh3 v1.0.2 // indirect
+	golang.org/x/mod v0.8.0 // indirect
+	golang.org/x/net v0.8.0 // indirect
+	golang.org/x/sys v0.6.0 // indirect
+	golang.org/x/text v0.8.0 // indirect
+	golang.org/x/tools v0.6.0 // indirect
+	golang.org/x/xerrors v0.0.0-20220609144429-65e65417b02f // indirect
+	google.golang.org/genproto v0.0.0-20230110181048-76db0878b65f // indirect
+	gopkg.in/yaml.v3 v3.0.1 // indirect
+)
```
Comment (HyukjinKwon) [inline go.mod:53]:
```suggestion
)

```

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,105 @@
+#
```
Comment (HyukjinKwon) [inline Makefile:1]: qq, did you write this Makefile? or get it from somewhere?

**Comment by HyukjinKwon** [comment]
Merged to master.

File: Makefile (hunk 4 above)
Comment (grundprinzip) [reply Makefile:1]: I wrote this.

**Comment by grundprinzip** [comment]
Thanks for merging

**Pull request closed without merging**